shapeshift/web
# fix: improve getUnderlyingAssetIdsBalances safety

## Description

See https://github.com/shapeshift/web/pull/4435 - fixes the root cause which that PR fixed with an opt chain, as well as other bugs previously fixed the same way.

The root issue was that we were early returning in case of no asset/underlyingAsset, which is wrong. Whether or not we have an asset/underlyingAsset for the precision, we should always return the same `GetUnderlyingAssetIdsBalancesReturn` i.e `Record<AssetId, { fiatAmount: string; cryptoBalancePrecision: string }>`.
tsc was somehow happy with the early returning of undefined, since this *technically* conformed to the types: whether or not there is an entry for the current `assetId`, the return type was still a valid `GetUnderlyingAssetIdsBalancesReturn`, but consumers actually expect to be able to access all underlyingAssetIds.

## Pull Request Type


## Issue (if applicable)

N/A

## Risk



If I got this wrong, this may bring back unsafety, sanity check consumptions. Did an LSP find references, and all consumptions should now be safe.

## Testing



- Do a smoke test of DeFi section - by provider/asset, expand and collapse rows. Ensure clicking opportunities is still working
- Ensure assets page is still displayed without errors, and clicking assets correctly routes to the asset page without any error - should be able to reload said asset page as well
- Ensure equity rows display without regressions
- Ensure Osmosis LP withdraw is still working and "Withdraw Continue" event still tracks underlying assets being withdrawn

### Engineering



### Operations



## Screenshots (if applicable)

<img width="1025" alt="image" src="https://user-images.githubusercontent.com/17035424/236424867-4a329985-f7e8-4673-904d-39437dd8ca60.png">
<img width="1043" alt="image" src="https://user-images.githubusercontent.com/17035424/236427496-fd6b48d4-74cf-4e0a-bf71-d4aef38210c8.png">
<img width="277" alt="image" src="https://user-images.githubusercontent.com/17035424/236428385-742f8484-b422-4083-9d07-9a02c008c5ea.png">
